implement "minimize or overview" with only current app windows

paroj opened the pull request:
or "Application Spread" as it was called back with Unity7.


https://user-images.githubusercontent.com/4941802/172016701-2032f8a2-2ebb-4f8e-b955-8347873a3661.mp4



---

This is a manual rebase of #519, which is by now almost exactly 3 years old. I also slightly reduced changes to `appIcons.js`. I hope it is enough for @micheleg to review this..

paroj commented:
@3v1n0

paroj commented:
@3v1n0 @micheleg

paroj commented:
@3v1n0 @micheleg

3v1n0 commented:
I've been giving a look to this, but it would need some extra efforts to support 3.38, sadly :-(

One quick change is the removal of the tweener.

3v1n0 commented on appExposeOverview.js:
```diff
@@ -0,0 +1,189 @@
+				w.x = startScaleX;
+				w.y = startScaleY;
+
+				Tweener.addTween(w, { time: 0.125, transition: "linear", scale_x: 1, scale_y: 1, x: originalX, y: originalY });
```
Please apply

```diff
diff --git a/appExposeOverview.js b/appExposeOverview.js
index 904c421..19dbf21 100644
--- a/appExposeOverview.js
+++ b/appExposeOverview.js
@@ -6,7 +6,6 @@ const Workspace = imports.ui.workspace;
 const WorkspaceThumbnail = imports.ui.workspaceThumbnail;
 const ViewSelector = imports.ui.viewSelector;
 const Layout = imports.ui.layout;
-const Tweener = imports.ui.tweener;
 
 const Me = imports.misc.extensionUtils.getCurrentExtension();
 const AppIcons = Me.imports.appIcons;
@@ -113,13 +112,13 @@ var AppExposeOverview = new Lang.Class({
 				Main.activateWindow(this.appWindows[0]);
 			}
 			// Add fade in when expose finished hiding,
-			// so non app expose windows dont suddenly pop up 
+			// so non app expose windows dont suddenly pop up
 			this._getNonExposeWindows().forEach(function (w) {
 				const originalX = w.x;
 				const originalY = w.y;
 
 				const monitorArea = global.display.get_monitor_geometry(w.get_meta_window().get_monitor())
-				
+
 				// version 1: start scaling non expose windows from monitor center
 				// const startScaleX = monitorArea.x + monitorArea.width/2;
 				// const startScaleY = monitorArea.y + monitorArea.height/2;
@@ -133,7 +132,14 @@ var AppExposeOverview = new Lang.Class({
 				w.x = startScaleX;
 				w.y = startScaleY;
 
-				Tweener.addTween(w, { time: 0.125, transition: "linear", scale_x: 1, scale_y: 1, x: originalX, y: originalY });
+				w.ease({
+					x: originalX,
+					y: originalY,
+					scale_x: 1,
+					scale_y: 1,
+					duration: 0.125 * 1000,
+					mode: Clutter.AnimationMode.LINEAR,
+				});
 			});
 		}
 	},
```

paroj commented:
done. Is there anything else? I am currently on Ubuntu 20.04 and cannot test with latest Gnome.
But you can to push to this branch on my repo directly if you find something.

paroj commented:
@3v1n0

paroj commented:
@3v1n0 @micheleg

micheleg commented:
I tried to run this on 3.38 but I get some errors:
```
JS ERROR: TypeError: windowActor.get_meta_window is not a function
show/this.destroyWindowId<@/home/michele/.local/share/gnome-shell/extensions/dash-to-dock@micxgx.gmail.com/appExposeOverview.js:70:35
_initializeUI/<@resource:///org/gnome/shell/ui/main.js:248:16
```

micheleg commented on appExposeOverview.js:
```diff
@@ -0,0 +1,195 @@
+		// Only show application windows in workspace
+		this.appWindows = appWindows;
+		const originalOverviewFunction = this.originalOverviewFunction;
+		Workspace.Workspace.prototype._isOverviewWindow = function(win) {
+			const originalResult = originalOverviewFunction(win);
+			const metaWindow = win.get_meta_window();
+			return originalResult && appWindows.indexOf(metaWindow) > -1;
+		};
```
```suggestion
		Workspace.Workspace.prototype._isOverviewWindow = function(metaWindow) {
			const originalResult = originalOverviewFunction(metaWindow);
			return originalResult && appWindows.indexOf(metaWindow) > -1;
		};
```

micheleg commented on appExposeOverview.js:
```diff
@@ -0,0 +1,195 @@
+
+		// Only show application windows in thumbnails
+		const originalThumbnailFunction = this.originalOverviewFunction;
+		WorkspaceThumbnail.WorkspaceThumbnail.prototype._isOverviewWindow = function(win) {
+			const originalResult = originalThumbnailFunction(win);
+			const metaWindow = win.get_meta_window();
+			return originalResult && appWindows.indexOf(metaWindow) > -1;
+		};
```
```suggestion
		WorkspaceThumbnail.WorkspaceThumbnail.prototype._isOverviewWindow = function(metaWindow) {
			const originalResult = originalThumbnailFunction(metaWindow);
			return originalResult && appWindows.indexOf(metaWindow) > -1;
		};
```

micheleg commented on appExposeOverview.js:
```diff
@@ -0,0 +1,195 @@
+		};
+
+		// If second last app window closed in AppExposeOverview, activate remaining window (done in hidden event)
+		this.destroyWindowId = global.window_manager.connect('destroy', (windowActor) => {
```
```suggestion
		this.destroyWindowId = global.window_manager.connect('destroy', (_, windowActor) => {
```

micheleg reviewed (commented):
Just enough to make it run on 3.38, not very thoughtful changes...

micheleg commented on appExposeOverview.js:
```diff
@@ -58,9 +58,8 @@ var AppExposeOverview = new Lang.Class({
-			const originalResult = originalThumbnailFunction(win);
-			const metaWindow = win.get_meta_window();
+		WorkspaceThumbnail.WorkspaceThumbnail.prototype._isOverviewWindow = function(metaWindow) {
+			const originalResult = originalThumbnailFunction(metaWindow);
```
This was actuallly right before and this breaks the thumbnails filter (not thumbnail shown), so it should be reverted.

paroj replied:
actually, I have no idea how `metaWindow` is different to `win` - I merely made some changes to the original code by @ggcode as far as I understood it.
Trying to learn about the GnomeShell API, hwoever, only [led me here](https://www.reddit.com/r/gnome/comments/f2qhrx/gnome_extensions_api/).

I just want to note, that you can directly push changes to my fork as you think fit. Also, maybe ggcode can help out.

3v1n0 commented:
Code changed quite a bit recently, would you mind rebasing, or should I try that?

paroj commented:
I would prefer you to do that, as i have no deeper understanding of the required API changes. Unless this is just about the git rebase operation.

ggcode commented:
@paroj @3v1n0 Have not been logged into this github account for over a year.. ;)
Was surprised to see the original code is still in discussion.
Just out of curiosity i fetched paroj's changes, (git) rebased it to micheleg's master and played around for an hour or so to (at least) get it working in > gnome 40 again:
https://github.com/micheleg/dash-to-dock/compare/master...ggcode:appspread-gnome40
- This is just enough to get it working in gnome 40. (i spreads the app windows again and you can select it in settings).
- Had to fake abort some unrelated? check in dash.js  (something with button.icon, no idea.. else it will crash).

ggcode commented:
@paroj Hi.. i rebased this branch (locally) and after that made the exact same changes i made here (changed more stuff compared to 4days ago): https://github.com/micheleg/dash-to-dock/compare/master...ggcode:appspread-gnome40
Before i push back into this branch:
- If you want to, fell free to test the new version .. and
- You maybe also want to save your current state into another branch (so, if somebody is interessted there will still be the pre gnome 40 version)?

paroj commented:
feel free to force-push here. I still have my changes locally.

Unfortunately, I wont be able to test with Gnome 40 until Ubuntu 22.04 arrives, as I am still on 20.04 & Gnome 36.

ggcode commented:
Rebase done.
Added new commit for the changes and refactoring for gnome  >= 40.

ggcode commented:
Regarding testing.. switched to Fedora on company laptop.. so i dont know / cant test on ubuntu <= gnome 40 ;)
Maybe @3v1n0 can try it out?
Last time i used virtualbox or somthing to test all versions from ~gnome 3.x, but im not using dash to dock all the time anyways.. 
So wont do that this time ;)

ggcode commented:
Needed to force push again as the default terminal git commit editor seems to be buggy and formated the commit message in a strange way.

paroj commented:
@3v1n0 can you take a look? It would be great to finally have this in Ubuntu 22.04 out of the box

3v1n0 commented:
I'm sorry I couldn't see this by Feature Freeze of 24.04, given that we were a bit overwhelmed of things... 
But well, we'll try to get this in by 24.10.

ggcode commented:
@paroj It's not enough to just resolve the conflict and merge.

Tried this (locally):
 Rebased current master. In master there is some new code reading searchbar dimensons in docking.js. But appspread-overview hides the searchbar and together with the master changes it doesnt work.

Created to versions that work:
1. Reenabled searchbar in appspread (tried to avoid that in the original version, but both overview and applications-grid use it, so why not appspread?)
2. Like current appspread branch (no searchbar in appspread, but using opacity instead of hiding). While trying that noticed one small glitch for one very rare usecase: If you switch directly from appspread to app-grid -> app-grid also has no search bar. Thats true for the actual version too. Could live with it but for discussion:

Added a commit, that has both (version 2 is just uncommenting this._disableSearch() && this._enableSearch()).

The merge conflict you mentioned could easily be resolved, but a rebase would be needed anyway (no time now, maybe will try later).

ggcode commented:
@paroj @3v1n0 Created a branch for you to easily test both versions 
* Switch versions in dtd settings-dialog
* Did some additional cleanup in v1 (without searchbar)
* It's also rebased to latest master
https://github.com/ggcode/dash-to-dock/tree/appspread_v1_v2
Please give it a try and let's decide which version to use.

ggcode commented:
Resolved merge conflict. Add forgoten gschema.xml.
Dont know whats the latest ubuntu branch to use.. as im now on gnome 42, tried it once against the latest tag. Worked.
https://github.com/ggcode/dash-to-dock/compare/ubuntu-dock-72_ubuntu5...ggcode:appspread_v1_v1-ubuntu-dock-72_ubuntu5

paroj commented:
squashed and rebased once more. Current state works with Ubuntu 22.04.

@3v1n0 you are aware of how spectacularly bad your project management is here, right?

paroj commented:
added a video to motivate merging this:

https://user-images.githubusercontent.com/4941802/172016701-2032f8a2-2ebb-4f8e-b955-8347873a3661.mp4

3v1n0 commented:
I've reworked on this but apparently I can't push commits to the original branch, so I've opened #1253, let me know if you can review it and rebase this branch.

Jules-Bertholet commented:
@3v1n0  I think you meant to link #1797 ?

paroj commented:
@3v1n0 I just tested #1797 and your changes LGTM.

I can force push them here or you just merge #1797

3v1n0 commented:
@paroj force push it here please, so we keep the history in a single place :)

3v1n0 commented:
I've rebased my branch on top of master, otherwise just tick the checkbox to allow me to push to your PR and I can handle it.

Cheers

paroj commented:
> I've rebased my branch on top of master, otherwise just tick the checkbox to allow me to push to your PR and I can handle it.

dang.. just did the same.. That checkbox was ticked all along. Maybe some permission/ group issue?

3v1n0 commented:
Uch I see a diff from my branch, please just `git fetch <my-remote>` and `git reset --hard 5a5d4`, then force-push again.

ggcode commented:
@paroj Thank you so much that we finally got this merged (5 years and 28 days later)
Just one sidenote: I did not try the latest version (yet). Have you tried that search (just starting to type chars on keyboard) is relly disabled while in app spread? Not sure from just reading the last version of the code. I tended to just dont enable/disable search in appspread. (less sources of errors in the future in case of gnome-shell changes)
But again: Thank you SO MUCH for all your effort and for finally making this happen. Its just a small addition, but im using it since 7 years now ;)

ggcode commented:
.. or i got spacing problems in gnome 43 with search opacity set to 0.. cant rember, just left search in appspread.
less code, didnt bother to see the search.. anyways thanks!

ggcode commented:
@paroj ok.. just cleaned up the extensions. installed latest dtd. reanabled appspread. works like a charm. might have been on the wrong branch regarding search concerns. And again: Really glad you keept urging to get that done. Thank you so much.

ggcode commented:
@paroj It was a long way since i was reading the comment 
"// TODO: only show current app windows when gnome shell API will allow it" in the original code of micheleg.
Now (according to omg ubuntu) i couldnt be happier that "Well, I’m pleased to say that request was merged and the change is default in Ubuntu 22.10."
Also thanks to @3v1n0 for cleaning up my code from 2017-2022 and making it a reality in ubuntu 22.10.

kacperpaczos commented:
@ggcode @3v1n0, do you think that enable spread function windows for default (like unity 7) is a good idea?

ggcode commented:
@kacperpaczos
As i tried to get that setting back for years (at least as option), i might be opinionated ;)
As i did not re-install a fresh version of dtd and not using ubuntu any more, im not even sure, that it is the default setting there (is it?)

For my (half shortcuts, half mouse) workflow its what feels most natrual to me.
* Got the dock on the left side of the primary monitor  (like ubuntu default, but autohide)
* If one window of an app is opened, it will directly hide or unhide that window.
* Else it will do the dance of show last openend window of the app if not active (switching from another app) OR
* (click dash icon again while app active) show/hide the spread windows only for the app clicked/choosen on dash (for me its easier/faster to find the exact window of the app in the larger previews there than lets say default gnome app switcher or the window previews setting in dash to dock).
* In Multimonitor setups its even easier to spot the exact window of the current app im looking for in appsread.
* I often have the usecase: multiple terminals, directories in nautilus/files, mutliple editors for different projects (vs-code or similar) etc..

I also assume 80% of the time most users dont even use two or more windows of the same app at the same time. So for them it doesnt even matter if its default or not.

To sum it up: Im ok with it beeing default. ;)

Glad you asked. Just curious: Why did you ask? Dont like it? Try to make it default somewhere else?

ggcode commented:
@kacperpaczos
Oh now i got it. Unity7 is the version used in the latest offical ubuntu unity flavor.
If that the question then: Yes!!, go for it as default :sunglasses: :thumbsup:

ryzendew commented:
Could you add an option like how cosmic dock does it also ? Nice work none the less
